File: src/main/java/dev/obscuria/tooltips/client/tooltip/element/QuadPalette.java
package dev.obscuria.tooltips.client.tooltip.element;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.tooltips.config.ARGBProvider;
import dev.obscuria.tooltips.util.color.ARGB;

@Desugar
public record QuadPalette(ARGBProvider topLeftProvider, ARGBProvider topRightProvider,
                          ARGBProvider bottomLeftProvider, ARGBProvider bottomRightProvider) {

    public static QuadPalette fromJson(JsonObject json) {
        return new QuadPalette(
                ARGBProvider.fromJson(json.get("top_left")),
                ARGBProvider.fromJson(json.get("top_right")),
                ARGBProvider.fromJson(json.get("bottom_left")),
                ARGBProvider.fromJson(json.get("bottom_right")));
    }

    public ARGB topLeft() {
        return this.topLeftProvider.get();
    }

    public ARGB topRight() {
        return this.topRightProvider.get();
    }

    public ARGB bottomLeft() {
        return this.bottomLeftProvider.get();
    }

    public ARGB bottomRight() {
        return this.bottomRightProvider.get();
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/style/QuadPalette.kt
package dev.obscuria.tooltips.client.style

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.fragmentum.tools.ARGB

class QuadPalette(
    val topLeft: ARGB,
    val topRight: ARGB,
    val bottomLeft: ARGB,
    val bottomRight: ARGB
) {

    companion object {

        val CODEC: Codec<QuadPalette> = RecordCodecBuilder.create { codec ->
            codec.group(
                ARGB.CODEC
                    .fieldOf("top_left")
                    .forGetter(QuadPalette::topLeft),
                ARGB.CODEC
                    .fieldOf("top_right")
                    .forGetter(QuadPalette::topRight),
                ARGB.CODEC
                    .fieldOf("bottom_left")
                    .forGetter(QuadPalette::bottomLeft),
                ARGB.CODEC
                    .fieldOf("bottom_right")
                    .forGetter(QuadPalette::bottomRight)
            ).apply(codec, ::QuadPalette)
        }
    }
}